In [2]:
#pip install gmaps

     |████████████████████████████████| 1.1 MB 2.5 MB/s eta 0:00:01


  Created wheel for gmaps: filename=gmaps-0.9.0-py2.py3-none-any.whl size=2076097 sha256=bae17f8fe7d3e76099d6d1e255c3a9d0542a4fd7019da972b375be8f3aab63ac
  Stored in directory: /Users/sunisapradhan/Library/Caches/pip/wheels/30/bd/87/d0d731ee712c34a4d6bace8b1dbf8c8932e22a1625b4a55be4
Successfully built gmaps
Note: you may need to restart the kernel to use updated packages.


### 6.5.2 Create Heatmaps for Weather Parameters 

In [1]:
# Import the dependencies.
import pandas as pd
import gmaps
import requests
# Import the API key.
from config import g_key

In [2]:
# Store the CSV you saved created in part one into a DataFrame.
city_data_df = pd.read_csv("weather_data/cities.csv")
city_data_df.head()

City_ID           City Country                 Date      Lat       Lng  \
0        0  Half Moon Bay      US  2021-08-09 21:22:54  37.4636 -122.4286   
1        1       Coquimbo      CL  2021-08-09 21:25:15 -29.9533  -71.3436   
2        2         Albany      US  2021-08-09 21:25:16  42.6001  -73.9662   
3        3       Hermanus      ZA  2021-08-09 21:25:16 -34.4187   19.2345   
4        4     Bairnsdale      AU  2021-08-09 21:25:17 -37.8333  147.6167   

   Max Temp  Humidity  Cloudiness  Wind Speed  
0     96.62        63           1        1.99  
1     58.73        67          20       11.50  
2     86.45        60          51        7.47  
3     49.89        71           3        4.14  
4     42.08        91           3        1.01

In [3]:
# Get the data types.
city_data_df.dtypes

City_ID         int64
City           object
Country        object
Date           object
Lat           float64
Lng           float64
Max Temp      float64
Humidity        int64
Cloudiness      int64
Wind Speed    float64
dtype: object

Create a Maximum Temperature Heatmap

- First, tell gmaps to use your API key. You only need to configure gmaps to use your API key once.

In [4]:
# Configure gmaps to use your Google API key.
gmaps.configure(api_key=g_key)

Next, create the heatmap for the maximum temperature. The general syntax for creating a heatmap is as follows.

In [5]:
# # 1. Assign the locations to an array of latitude and longitude pairs.
# locations = [latitude, longitude]
# # 2. Assign the weights variable to some values.
# temperatures = # an array of length equal to the locations array length
# # 3. Assign the figure variable to the gmaps.figure() attribute.
# fig = gmaps.figure()
# # 4. Assign the heatmap_layer variable to the heatmap_layer attribute and add in the locations.
# heatmap_layer = gmaps.heatmap_layer(locations, weights=temperatures)

# # 5. Add the heatmap layer.
# fig.add_layer(heatmap_layer)
# # 6. Call the figure to plot the data.
# fig

In [10]:
# Get the maximum temperature.
max_temp = city_data_df["Max Temp"]
temps = []
for temp in max_temp:
    temps.append(max(temp, 0))

Instead of using the for loop, we can perform a list comprehension within the heatmap_layer() function.

Replace temps with our code for the list comprehension so our heat_layer code looks like the following:

heat_layer = gmaps.heatmap_layer(locations, weights=[max(temp, 0) for temp in max_temp])

In [16]:
# Heatmap of temperature
# Get the latitude and longitude.
locations = city_data_df[["Lat", "Lng"]]
# Get the maximum temperature.
max_temp = city_data_df["Max Temp"]
# Assign the figure variable.
fig = gmaps.figure(center=(30.0, 31.0), zoom_level=1.5)
# Assign the heatmap variable.
#heat_layer = gmaps.heatmap_layer(locations, weights=max_temp)
#heat_layer = gmaps.heatmap_layer(locations, weights=temps, dissipating=False, max_intensity=300, point_radius=4)
#heat_layer = gmaps.heatmap_layer(locations, weights=[max(temp, 0) for temp in max_temp])
# Add the heatmap layer.
fig.add_layer(heat_layer)
# Call the figure to plot the data.
fig

Figure(layout=FigureLayout(height='420px'))

This map is too large. We will need to make some adjustments to the gmaps.figure() attribute.

### Adjust Heatmap Zoom, Intensity, and Point Radius 

First, add the geographic center of Earth in the form of latitude and longitude (30.0° N and 31.0° E). Also, add a zoom level so that only one map of Earth is shown.

When we add a center and zoom level to the gmaps.figure() attribute, it will look like this:



### Create a Percent Humidity Heatmap 

In [19]:
# Heatmap of percent humidity
locations = city_data_df[["Lat", "Lng"]]
humidity = city_data_df["Humidity"]
fig = gmaps.figure(center=(30.0, 31.0), zoom_level=1.5)
heat_layer = gmaps.heatmap_layer(locations, weights=humidity, dissipating=False, max_intensity=300, point_radius=4)

fig.add_layer(heat_layer)
# Call the figure to plot the data.
fig

Figure(layout=FigureLayout(height='420px'))

### Create a % Cloudiness Heatmap 

In [20]:
# Heatmap of percent cloudiness
locations = city_data_df[["Lat", "Lng"]]
clouds = city_data_df["Cloudiness"]
fig = gmaps.figure(center=(30.0, 31.0), zoom_level=1.5)
heat_layer = gmaps.heatmap_layer(locations, weights=clouds, dissipating=False, max_intensity=300, point_radius=4)

fig.add_layer(heat_layer)
# Call the figure to plot the data.
fig

Figure(layout=FigureLayout(height='420px'))

### Create a Wind Speed Heatmap 

In [22]:
# Heatmap of percent humidity
locations = city_data_df[["Lat", "Lng"]]
wind = city_data_df["Wind Speed"]
fig = gmaps.figure(center=(30.0, 31.0), zoom_level=1.5)
heat_layer = gmaps.heatmap_layer(locations, weights=wind, dissipating=False, max_intensity=300, point_radius=4)

fig.add_layer(heat_layer)
# Call the figure to plot the data.
fig

Figure(layout=FigureLayout(height='420px'))

## 6.5.3 Get Vacation Criteria 

For the app we are creating, we need to prompt the user to enter the minimum and maximum temperature ranges as floating-point decimal numbers to filter the city_data_df DataFrame.
- Recall that you use the input() statement to prompt a user to enter information. To convert the input to a floating-point decimal number, wrap the input() statement with the float() method, like this: float(input()).


In [24]:
# Ask the customer to add a min and max temp value
min_temp = float(input("What is the minimum temperature you would like for your trip? "))
max_temp = float(input("What is the maximum temperature you would like for your trip? "))

What is the minimum temperature you would like for your trip? 75
What is the maximum temperature you would like for your trip? 90


Next, filter the maximum temperature column in the city_data_df DataFrame using logical operators to create a new DataFrame with the cities that meet the customer's criteria.
- Recall that you use the loc[] method on the current DataFrame to create a new DataFrame from a current DataFrame. Inside the brackets, add the conditional filter.


In [25]:
# Filter the dataset to find the cities that fit the criteria.
preferred_cities_df = city_data_df.loc[(city_data_df["Max Temp"] <= max_temp) & \
                                       (city_data_df["Max Temp"] >= min_temp)]
preferred_cities_df.head(10)

City_ID        City Country                 Date      Lat       Lng  \
2         2      Albany      US  2021-08-09 21:25:16  42.6001  -73.9662   
6         6      Atuona      PF  2021-08-09 21:25:17  -9.8000 -139.0333   
9         9     Ransang      PH  2021-08-09 21:25:19   8.9603  117.5828   
10       10  Coahuayana      MX  2021-08-09 21:25:19  18.7333 -103.6833   
12       12  Butaritari      KI  2021-08-09 21:25:20   3.0707  172.7902   
14       14  Sao Filipe      CV  2021-08-09 21:25:20  14.8961  -24.4956   
15       15   Jaisalmer      IN  2021-08-09 21:25:20  26.9147   70.9181   
16       16     Da Nang      VN  2021-08-09 21:25:21  16.0678  108.2208   
19       19        Hilo      US  2021-08-09 21:21:26  19.7297 -155.0900   
25       25   Jamestown      US  2021-08-09 21:22:01  42.0970  -79.2353   

    Max Temp  Humidity  Cloudiness  Wind Speed  
2      86.45        60          51        7.47  
6      78.04        68          14       19.24  
9      78.85        85         100        4.03  
10     87.28        78         100        9.82  
12     82.92        66          16       13.94  
14     78.55        86          52        8.57  
15     83.01        61           1       21.72  
16     86.05        83          20        1.14  
19     83.25        68           1        8.05  
25     89.02        49           1        3.00

In [27]:
# Determine if there are any null values in the preferred_cities_df DataFrame:
# preferred_cities_df.count()
# preferred_cities_df.isnull().sum()
# preferred_cities_df.notnull().sum()

preferred_cities_df.count()

City_ID       191
City          191
Country       191
Date          191
Lat           191
Lng           191
Max Temp      191
Humidity      191
Cloudiness    191
Wind Speed    191
dtype: int64

For the above DataFrame, there are 191 cities and no null values

- Depending on the time of year and the seasons, you might have to adjust the minimum and maximum temperature to get enough cities.
- It is a good idea to keep the number of cities to fewer than 200 to make it easier to plot the markers on the heatmap.
- If you have some rows with null values, you'll need to drop them using the dropna() method at the end of your filtering statement when you are creating the new DataFrame.

## 6.5.4 Map Vacation Criteria 

Once the customers have filtered the database (DataFrame) based on their temperature preferences, show them a heatmap for the maximum temperature for the filtered cities. In addition, create a marker for each city that will display the name of the city, country code, maximum temperature, and name of a nearby hotel within three miles of the coordinates when the marker is clicked.

Using the coordinates from the preferred_cities_df DataFrame, find a hotel using our Google Places API and then retrieve that hotel information. Once we retrieve the hotel information, we'll need to store it so we can reference it and add the information to the pop-up marker.



- Don't add the hotel information to the preferred_cities_df DataFrame because this DataFrame is our filtered DataFrame, and the customer will always filter it for each trip. We'll need to create a new DataFrame specifically for the data needed to create a heatmap and pop-up markers.

- Make a copy of the preferred_cities_df DataFrame and name it hotel_df. For the hotel_df, keep the columns "City," "Country," "Max Temp," "Lat," and "Lng." Add a new column to the hotel_df DataFrame to hold the name of the hotel.

- Add the following code to a new cell and run it to create the hotel_df DataFrame:
hotel_df = preferred_cities_df[["City", "Country", "Max Temp", "Lat", "Lng"]].copy()
hotel_df["Hotel Name"] = ""
hotel_df.head(10)

In [29]:
# Create DataFrame called hotel_df to store hotel names along with city, country, max temp, and coordinates.
hotel_df = preferred_cities_df[["City", "Country", "Max Temp", "Lat", "Lng"]].copy()
hotel_df["Hotel Name"] = ""
hotel_df.head(10)

City Country  Max Temp      Lat       Lng Hotel Name
2       Albany      US     86.45  42.6001  -73.9662           
6       Atuona      PF     78.04  -9.8000 -139.0333           
9      Ransang      PH     78.85   8.9603  117.5828           
10  Coahuayana      MX     87.28  18.7333 -103.6833           
12  Butaritari      KI     82.92   3.0707  172.7902           
14  Sao Filipe      CV     78.55  14.8961  -24.4956           
15   Jaisalmer      IN     83.01  26.9147   70.9181           
16     Da Nang      VN     86.05  16.0678  108.2208           
19        Hilo      US     83.25  19.7297 -155.0900           
25   Jamestown      US     89.02  42.0970  -79.2353

Using the latitude and longitude and specific parameters, use the Google Places Nearby Search request to retrieve a hotel and add it to the Hotel Name column.
## Retrieve Hotels from a Nearby Search
The first step for retrieving hotels from a Nearby Search is to set the parameters for the search.

In [33]:
# Set the parameters to search for a hotel in Paris.
params = {
    "radius": 5000,
    "types": "lodging",
    "key": g_key,
    "location": "48.8566, 2.3522"}
# Use base URL to search for hotels in Paris.
base_url = "https://maps.googleapis.com/maps/api/place/nearbysearch/json"
# Make request and get the JSON data from the search.
hotels = requests.get(base_url, params=params).json()

hotels

{'html_attributions': [],
 'next_page_token': 'Aap_uEArKXytwpZcEKyagcEcVQlB4Fj53NYRvAfqbF0DV0qtp22U9eMu8cWbMbB-VjF0XN__0-BdxUADhZA83SnkioKqP2Y4apuoXisMYFK7xmacjr8sOn8h-Fo92PQM34iobK_li9-UI60tB08GBJV0CBhbYY27HMnngMcM0eFglAjtJbOJdYPZVIRjSMnMRLM4ZnWICjgTft1LhZ74_xVxvoG6gdeMNTj1y2iTneC51hb32ylQ0nEsmYMa21LK1TO4UReNLvFih_zEmddhz_RE90kLqZLXNfEaCd-_aHMsRW_G--zECwm2-yTSwV9LxFYydcE-M8LFUY1aYg-X6qp--pbE-vFELNPnzPY1NPrKCKcp3Zmmn_4s2rJVgin4cjNMv22Uikv-2be7OW6ZiwOgCLeGXPmjcP_6DIx07AIf2j-CO50Cieem2JdE',
 'results': [{'business_status': 'OPERATIONAL',
   'geometry': {'location': {'lat': 48.8581126, 'lng': 2.3529277},
    'viewport': {'northeast': {'lat': 48.8594393802915,
      'lng': 2.354352980291503},
     'southwest': {'lat': 48.8567414197085, 'lng': 2.351655019708499}}},
   'icon': 'https://maps.gstatic.com/mapfiles/place_api/icons/v1/png_71/lodging-71.png',
   'icon_background_color': '#909CE1',
   'icon_mask_base_uri': 'https://maps.gstatic.com/mapfiles/place_api/icons/v2/hotel_pinlet',
   'nam

## Iterate Thru hotel_df DataFrame 

We can use the iterrows() function to perform the iteration; however, we need to provide the index and the row in the for loop using this syntax: for index, row in df.iterrows().

In [36]:
# Iterate through the DataFrame.
for index, row in hotel_df.iterrows():
    # Get the latitude and longitude.
    lat = row["Lat"]
    lng = row["Lng"]

    # Add the latitude and longitude to location key for the params dictionary.
    params["location"] = f"{lat},{lng}"

    # Use the search term: "lodging" and our latitude and longitude.
    base_url = "https://maps.googleapis.com/maps/api/place/nearbysearch/json"
    # Make request and get the JSON data from the search.
    hotels = requests.get(base_url, params=params).json()
    # Grab the first hotel from the results and store the name.
    try:
        hotel_df.loc[index, "Hotel Name"] = hotels["results"][0]["name"]
    except (IndexError):
        print("Hotel not found... skipping.")


Hotel not found... skipping.
Hotel not found... skipping.
Hotel not found... skipping.
Hotel not found... skipping.
Hotel not found... skipping.
Hotel not found... skipping.
Hotel not found... skipping.
Hotel not found... skipping.
Hotel not found... skipping.
Hotel not found... skipping.
Hotel not found... skipping.
Hotel not found... skipping.
Hotel not found... skipping.
Hotel not found... skipping.


In [39]:
hotel_df.head(10)

City Country  Max Temp      Lat       Lng  \
2       Albany      US     86.45  42.6001  -73.9662   
6       Atuona      PF     78.04  -9.8000 -139.0333   
9      Ransang      PH     78.85   8.9603  117.5828   
10  Coahuayana      MX     87.28  18.7333 -103.6833   
12  Butaritari      KI     82.92   3.0707  172.7902   
14  Sao Filipe      CV     78.55  14.8961  -24.4956   
15   Jaisalmer      IN     83.01  26.9147   70.9181   
16     Da Nang      VN     86.05  16.0678  108.2208   
19        Hilo      US     83.25  19.7297 -155.0900   
25   Jamestown      US     89.02  42.0970  -79.2353   

                           Hotel Name  
2                                      
6                         Villa Enata  
9     Black and Crystal Sand Campsite  
10  El rranchito michuacan de ocampos  
12                 Isles Sunset Lodge  
14                      Hotel Xaguate  
15                         Hotel Fifu  
16                    Sun River Hotel  
19                Hilo Hawaiian Hotel  
25     Hampton Inn & Suites Jamestown

If you encounter more errors, add them to the except block—like so, (IndexError, NewError)—and continue running your code until there are no errors.

## Create a Max Temp Heatmap from a Hotel DataFrame 

Before we add pop-up markers with hotel information, let's create a heatmap using the maximum temperature from our hotel_df DataFrame by reusing the code and changing the DataFrame name.

In [37]:
# Add a heatmap of temperature for the vacation spots.
locations = hotel_df[["Lat", "Lng"]]
max_temp = hotel_df["Max Temp"]
fig = gmaps.figure(center=(30.0, 31.0), zoom_level=1.5)
heat_layer = gmaps.heatmap_layer(locations, weights=max_temp, dissipating=False,
             max_intensity=300, point_radius=4)

fig.add_layer(heat_layer)
# Call the figure to plot the data.
fig

Figure(layout=FigureLayout(height='420px'))

Let's add the markers for each city on top of the heatmap

In [38]:
# Add a heatmap of temperature for the vacation spots.
locations = hotel_df[["Lat", "Lng"]]
max_temp = hotel_df["Max Temp"]
fig = gmaps.figure(center=(30.0, 31.0), zoom_level=1.5)
heat_layer = gmaps.heatmap_layer(locations, weights=max_temp, dissipating=False,
             max_intensity=300, point_radius=4)
marker_layer = gmaps.marker_layer(locations)
fig.add_layer(heat_layer)
fig.add_layer(marker_layer)
# Call the figure to plot the data.
fig

Figure(layout=FigureLayout(height='420px'))

In [41]:
info_box_template = """
<dl>
<dt>Hotel Name</dt><dd>{Hotel Name}</dd>
<dt>City</dt><dd>{City}</dd>
<dt>Country</dt><dd>{Country}</dd>
<dt>Max Temp</dt><dd>{Max Temp} °F</dd>
</dl>
"""

# Store the DataFrame Row.
hotel_info = [info_box_template.format(**row) for index, row in hotel_df.iterrows()]

In [42]:
# Add a heatmap of temperature for the vacation spots and a pop-up marker for each city.
locations = hotel_df[["Lat", "Lng"]]
max_temp = hotel_df["Max Temp"]
fig = gmaps.figure(center=(30.0, 31.0), zoom_level=1.5)
heat_layer = gmaps.heatmap_layer(locations, weights=max_temp,dissipating=False,
             max_intensity=300, point_radius=4)
marker_layer = gmaps.marker_layer(locations, info_box_content=hotel_info)
fig.add_layer(heat_layer)
fig.add_layer(marker_layer)

# Call the figure to plot the data.
fig

Figure(layout=FigureLayout(height='420px'))